Exercises/multi_gpu_cuda/Solution/multi_gpu_cuda.cpp

```diff
@@ -0,0 +1,189 @@
+//@HEADER
+// ************************************************************************
+//
+//                        Kokkos v. 4.0
+//       Copyright (2022) National Technology & Engineering
+//               Solutions of Sandia, LLC (NTESS).
+//
+// Under the terms of Contract DE-NA0003525 with NTESS,
+// the U.S. Government retains certain rights in this software.
+//
+// Part of Kokkos, under the Apache License v2.0 with LLVM Exceptions.
+// See https://kokkos.org/LICENSE for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//@HEADER
+
+#include <Kokkos_Core.hpp>
+
+#ifndef KOKKOS_ENABLE_CUDA
+#error "This exercise can only be run with Kokkos_ENABLE_CUDA=ON"
+#else
+
+using HostSpace      = Kokkos::HostSpace;
+using ExecSpace      = Kokkos::DefaultExecutionSpace;
+using TeamPolicy     = Kokkos::TeamPolicy<ExecSpace>;
+using MemberType     = TeamPolicy::member_type;
+using ViewVectorType = Kokkos::View<double*>;
+using ViewMatrixType = Kokkos::View<double**>;
+
+// Use a view result type for parallel reduce to avoid fencing over all
+// execution spaces. Without this the kernels will run serially.
+using ResultType = Kokkos::View<double>;
+
+struct CudaStreams {
+  std::array<int, 2> devices;
+  std::array<cudaStream_t, 2> streams;
+
+  CudaStreams() {
+    // Query number of devices available
+    int n_devices;
+    KOKKOS_IMPL_CUDA_SAFE_CALL(cudaGetDeviceCount(&n_devices));
```

Comment:
Let's not showcase using private macros.

Reply:
Removed.